src/shared/serialize.c

```diff
@@ -324,6 +324,42 @@ int deserialize_environment(const char *value, char ***list) {
         return 0;
 }
 
+int deserialize_fd(const char *value, FDSet *fds, bool store_index) {
+        int fd, r;
+
+        assert(value);
+        assert(fds);
+
+        if (!store_index) {
+                fd = parse_fd(value);
+                if (fd < 0)
+                        return log_debug_errno(fd, "Failed to parse FD out of value: %s", value);
+
+                if (!fdset_contains(fds, fd))
+                        return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "FD %d not in fdset.", fd);
+
+                r = fdset_remove(fds, fd);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to remove value=%d from fdset", fd);
+        } else {
+                size_t i;
+
+                r = safe_atozu(value, &i);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to parse FD index out of value: %s", value);
+
+                if (i >= fdset_size(fds))
+                        return log_debug_errno(SYNTHETIC_ERRNO(ERANGE), "FD index %zu not in fd array.", i);
+
+                /* Can't remove from the set, or the order will change */
+                FDSET_FOREACH(fd, fds)
```

Comment:
ahum. this is terrible. the sets might be huge...

also, should be in some fdset-specific accessor function

Reply:
i don't have a nice way to speed this up, but linear counting sounds awful...

Reply:
ok, fixed by converting the set to an array in the deserializing helper, at the beginning, so that if the serialization is by index it is used instead of the set